cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -442,16 +442,18 @@ newtype TPosition =
 
 private newtype TReturnKind =
   TNormalReturnKind(int indirectionIndex) {
-    Ssa::hasIndirectOperand(any(ReturnValueInstruction ret).getReturnAddressOperand(),
-      indirectionIndex + 1) // `indirectionIndex` is one less than the argument because the return loads the value.
-    or
-    indirectionIndex = 0 // TODO: very much a bodge so that it works on the test that has no return statements
+    indirectionIndex =
+      [0 .. max(Ssa::Function f |
+          |
+          Ssa::getMaxIndirectionsForType(f.getUnspecifiedType()) - 1 // -1 because a returned value is a prvalue not a glvalue
+        )]
```

Comment:
So instead of relying on the presence of a return value instruction we're now relying on the function's return type? I can see how that also works, but I have some concerns:
- This makes dataflow more dependent on the surface syntax in a way that it wasn't before. I'd prefer to keep dataflow completely unaware of the C/C++ AST (and with a very limited set of cases it is) since it means we can freely change AST classes in the future, and as long as the IR stays correct dataflow will do the right thing.
An example of where this may bite us: In the IR we synthesize function bodies for some things that are not `Function`s. In particular, we do so for the initializer of a global variable and for `static` local variables.
- Eventually, I'd like to make the construction of indirection nodes smarter to handle things like:
  ```cpp
  unsigned long get_data_as_long() {
    char* my_data = source(); // *my_data is tainted
    unsigned long my_data_as_an_unsigned_long = (unsigned long)my_data; // *my_data_as_an_unsigned_long should be tainted, but unsigned longs have no indirection nodes. So flow is lost!
    return my_data_as_an_unsigned_long;
  }

  unsigned long data = my_data_as_an_unsigned_long();
  char* my_data2 = (char*)data;
  sink(my_data2);
  ```
  The solution to the above will be to be smarter about how many dataflow nodes we allocate (see [this internal link](https://github.com/github/codeql-c-team/issues/2062) for more information on this).

  And since the current logic in the `TNormalReturnKind` branch only talks about the presence of an indirection operand that logic will continue to work once allocate indirection nodes in a smarter way. In contrast, the change here hardwires that we create return kinds based on the return type of a function.

Instead, I think a better solution would be:
- Keep the existing logic
- Replace the `indirectionIndex = 0 // TODO: very much a bodge so that it works on the test that has no return statements` disjunct with a call into the flow summary library (i.e., the library you're filling in in this PR) that finds the set of indirections necessary for the modelled functions that exists as MaD rows. That is, we should have a predicate in `FlowSummaryImpl` that reads something like:
```qll
/**
 * Gets the maximum number of indirections that can be returned by a the function
 * modelled using the MaD row `package;type;subtypes;name;signature;ext`.
 */
int maxIndirectionForModelledFunction(string package, string type, boolean subtypes, string name, string signature, string ext) {
  exists(interpretElement(package, type, subtypes, name, signature, ext)) and
  result = /* Extract the number of stars in the return type specified by signature */
}
```
and then we could call this predicate instead of hardcoding the `indirectionIndex = 0` case. i.e., we'd do something like:
```ql
indirectionIndex = maxIndirectionForModelledFunction(_, _, _, _, _, _)
```
this would ensure that we have `ReturnKind`s for all the MaD rows we have, and we keep the nice property that we don't depend on the C++ surface AST for most of dataflow.

Reply:
Makes sense, though notably the original code doesn't even work any more - I'm trying to debug a monotonic recursion error with that, *then* wire up the new case.

Reply:
I'm not sure, I've rebased this PR a couple of times so the history is quite confusing.

The issue appears to be that `TNormalReturnKind` depends on `DataFlow::Node` in order to find all the return kinds / indirection levels; and `DataFlow::Node` now includes `TFlowSummaryNode`, i.e. depends on the flow summary library, and *that* of course depends on return kinds.  Theoretically we can break the loop by having `TNormalReturnKind` only depend on the non-summary `DataFlow::Node`s (and having a separate check of the flow summaries themselves as you suggest) ... but simply adding `not return instanceof TFlowSummaryNode` in `TNormalReturnKind` does not actually break the dependency.

Reply:
... I'm coming back to the idea that return kinds should be lower level than data flow nodes, and thus should not depend on them.  This way its easy to reason about them and it doesn't produce any dependency cycles.  If we do smarter things with indirection nodes in the future, we'll just have to extend the `ReturnKind` type manually as part of that process.

The above means leaving the code more-or-less how it currently is in this PR.  It might be worth adding the `maxIndirectionForModelledFunction` thing as well.  Is there actually any alternative design?

Reply:
> but simply adding `not return instanceof TFlowSummaryNode` in TNormalReturnKind does not actually break the dependency.

Indeed, that won't break any dependencies.

I'll pull your PR down locally and see if I can see what's wrong. The changes we did very early in this PR should've prevented exactly this non-monotonic recursion.

Reply:
I think you've just gotten a bit lost in the commit history 😂 Doing the changes in https://github.com/github/codeql/pull/15371/commits/2bea0adb9218b900bd149b7fdca03e91823c87be solves the non-monotonic recursion problem. When you said:
> Makes sense, though notably the original code doesn't even work any more - I'm trying to debug a monotonic recursion error with that, then wire up the new case.

I guess you mean that the code on `main` doesn't work any more. That's indeed expected - and that's why we did https://github.com/github/codeql/pull/15371/commits/2bea0adb9218b900bd149b7fdca03e91823c87be. Reintroducing those changes fixed the non-motonic recursion

Reply:
> ... I'm coming back to the idea that return kinds should be lower level than data flow nodes, and thus should not depend on them. This way its easy to reason about them and it doesn't produce any dependency cycles.

I totally agree. That's what https://github.com/github/codeql/commit/2bea0adb9218b900bd149b7fdca03e91823c87be did. It ensured that `ReturnKind`s didn't depend on dataflow nodes, but only on SSA.

Reply:
I've pushed two commits now:
- https://github.com/github/codeql/pull/15371/commits/21592563cb1b6e9bcb9bd5918fa9c6268c2a024d reintroduces the changes from https://github.com/github/codeql/commit/2bea0adb9218b900bd149b7fdca03e91823c87be (This is now slightly simpler because we don't need the `parameterIsRedefined` predicate since it was deleted from the code on `main` 🎉).
- https://github.com/github/codeql/pull/15371/commits/448a9015c05e4b850f6e8f3543b09ebe49613b8c adds a predicate to compute the indirections necessary from all the MaD models like I described in https://github.com/github/codeql/pull/15371#discussion_r1474890950